backend/src/database/InMemoryDatasource.ts

```diff
@@ -0,0 +1,84 @@
+import { DataSource, DataSourceConfig } from 'apollo-datasource';
+import { Post } from '../interfaces/Post';
+import { User } from '../interfaces/User';
+import { MutationResolvers } from '../generated/graphqlgen';
+import { randomBytes } from 'crypto';
+import PostInput = MutationResolvers.PostInput;
+import UserInput = MutationResolvers.UserInput;
+import { Context } from "../types";
+
+export class InMemoryDatasource extends DataSource {
+    users: User[] = [];
+
+    posts: Post[] = [];
+
+    constructor() {
+        super();
+    }
+
+    initialize(config: DataSourceConfig<Context>): void | Promise<void> {
+        //return super.initialize(config);
+    }
+
+    set Posts(posts: Post[]) {
+        this.posts = posts;
+    }
+
+    set Users(users: User[]) {
+        this.users = users;
+    }
+
+    getAllPosts(): Post[] {
+        return this.posts;
+    }
+
+    getAllUsers(): User[] {
+        return this.users;
+    }
+
+    getAllPostsOfOneUser(name: string): Post[] {
+        return this.posts.filter(e => e.author.name === name);
+    }
+
+    getAuthor(id: string): User {
+        return this.posts.filter(e => e.id === id)[0].author;
+    }
+
+    upvotePost(id: string, voter: UserInput): Post {
+        return this.votePost(id, voter.name, true);
+    }
+
+    downVotePost(id: string, voter: UserInput): Post {
+        return this.votePost(id, voter.name, true);
+    }
+
+    writePost(post: PostInput): Post {
+        const newPost: Post = {
+            ...post,
+            id: randomBytes(16).toString('hex'),
```

Comment:
This would be true as along as you do not generate the ID's etc based on an actual database. Because a constructor should never fail and the generator function can fail (if the database is unavailable, id out of bounds, ...).